cmd/generate-catalog/types.go

```diff
@@ -0,0 +1,262 @@
+	}
+}
+
+// addDeprecations adds a "olm.deprecations" object to the base catalog.
+func (c *CatalogTemplate) addDeprecations(deprecations Deprecations) {
+	c.Entries = append(c.Entries, CatalogEntry(deprecations))
+}
+
+// addBundles adds a list of "olm.bundle" objects to the base catalog.
+func (c *CatalogTemplate) addBundles(bundles []BundleEntry) {
+	for _, bundle := range bundles {
+		c.Entries = append(c.Entries, CatalogEntry(bundle))
+	}
+}
+
+// Create a new "olm.channel" object which should be added to the catalog base.
+// it will be represented in YAML like this:
+// |  - schema: olm.channel
+// |    name: rhacs-3.64
+// |    package: rhacs-operator
+// |    entries:
+// |      - <ChannelEntry>
+func newChannel(version *semver.Version, entries []ChannelEntry) *Channel {
+	return &Channel{
+		Schema:  "olm.channel",
+		Name:    generateChannelName(version),
+		Package: "rhacs-operator",
+		Entries: entries,
+	}
+}
+
+// Create a special "olm.channel" object with name "latest".
+// It is a deprecated channel which was used before 4.x.x version.
+func newLatestChannel(entries []ChannelEntry) Channel {
+	return Channel{
+		Schema:  "olm.channel",
+		Name:    "latest",
+		Package: "rhacs-operator",
+		Entries: entries,
+	}
+}
+
+// Create a special "olm.channel" object with name "stable".
+// It is a default channel for all versions after 4.x.x.
+func newStableChannel(entries []ChannelEntry) Channel {
+	return Channel{
+		Schema:  "olm.channel",
+		Name:    "stable",
+		Package: "rhacs-operator",
+		Entries: entries,
+	}
+}
+
+// Create a new Chanel entry object which should be added to Channel entries list.
+// it will be represented in YAML like this:
+// |  - name: rhacs-operator.v<version>
+// |    replaces: rhacs-operator.v<previousEntryVersion>
+// |    skipRange: '>= <previousChannelVersion> < <version>'
+// |    skips:
+// |      - rhacs-operator.v<broken_version>
+func newChannelEntry(version, previousEntryVersion, previousChannelVersion *semver.Version, brokenVersions []*semver.Version) ChannelEntry {
+	entry := ChannelEntry{
+		Name: generateBundleName(version),
+	}
+	entry.addReplaces(version, previousEntryVersion)
+	entry.addSkipRange(version, previousChannelVersion)
+	entry.addSkips(version, brokenVersions)
+	return entry
+}
+
+func (e *ChannelEntry) addReplaces(version, previousEntryVersion *semver.Version) {
+	// skip setting "replaces" key for the first version in each major version
+	versionsWithoutReplaces := []string{first3MajorVersion, first4MajorVersion}
+
+	replacesVersion := previousEntryVersion.Original()
+	if !slices.Contains(versionsWithoutReplaces, version.Original()) {
+		e.Replaces = "rhacs-operator.v" + replacesVersion
+	}
+}
+
+func (e *ChannelEntry) addSkipRange(version, previousChannelVersion *semver.Version) {
+	skipRangeGreaterThanEqual := fmt.Sprintf("%d.%d.0", previousChannelVersion.Major(), previousChannelVersion.Minor())
+	skipRangeLessThan := version.Original()
```

Comment:
Optional renaming suggestion:
```suggestion
	skipRangeFrom := fmt.Sprintf("%d.%d.0", previousChannelVersion.Major(), previousChannelVersion.Minor())
	skipRangeTo := version.Original()
```